rust-symcrypt/src/hmac.rs

```diff
@@ -1,180 +1,419 @@
+/* Hmac functions. For further documentation please refer to symcrypt.h */
+
+use crate::errors::SymCryptError;
 use core::ffi::c_void;
 use std::mem;
+use std::pin::Pin;
 use std::ptr;
+use std::sync::Arc;
 use symcrypt_sys;
 
 pub const SHA256_HMAC_RESULT_SIZE: usize = symcrypt_sys::SYMCRYPT_SHA256_RESULT_SIZE as usize;
 pub const SHA384_HMAC_RESULT_SIZE: usize = symcrypt_sys::SYMCRYPT_SHA384_RESULT_SIZE as usize;
 
+/// Generic trait for stateful Hmac functions
+///
+/// append() appends data to the HmacShaXXXState, this operation can be done multiple times.
+///
+/// result() returns the result of the Hmac. Once result() is called, the lifetime of the HmacShaXXXState is
+/// finished and HmacShaXXState will be drop()'d. To perform other stateful hash operation you must create
+/// a new hash object via HmacShaXXXState::new()
+///
+/// copy() creates a copy of the current HmacShaXXXState. copy() will create a new state but will reference the same
+/// expanded_key of the current HmacShaXXXState; therefore increasing the refcount of the HmacShaXXXExpandedKey struct.
+pub trait HmacState {
+    type Result;
+
+    fn append(&mut self, data: &[u8]);
+    fn result(self) -> Self::Result;
+    fn copy(&self) -> Box<Self>;
+}
+
+/// Wrapper for the expanded key that is Pin<Box<>>'d because we want to ensure that its memory address is heap allocated and not moved through its lifetime.
+struct HmacSha256ExpandedKey(Pin<Box<symcrypt_sys::SYMCRYPT_HMAC_SHA256_EXPANDED_KEY>>);
```

Comment:
@eholk  is `Pin<Box<>>` overkill here? Im assuming `Arc<>` must have some sort of built in Boxing? I think we'd rather be safe than sorry here with SymCrypt's memory managment stuff

Reply:
The `Pin<Box<...>>` is overkill here. Instead, I'd change line 60 to say `expanded_key: Pin<Arc<HmacSha256ExpandedKey>>` (I'll leave a comment there too).

Both `Box` and `Arc` are heap allocations. The difference is that `Box` is uniquely owned, while `Arc` can have shared ownership. `Box` is like `unique_ptr` in C++ and `Arc` is like `shared_ptr`.

Reply:
However, this needs to remain a Box so it remains as a heap pointer.

Reply:
The reason we have `Pin<Box<>>` here is because in the `new()` we first initialize the variable `expanded_key` on line 73 and we are worried that rust might move the memory address before we are able to `Pin<Arc<>>` on line 86,

Reply:
I will try to remove the `Pin<Box<>>` and see if there is any errors in the unit test